upup/pkg/fi/cloudup/scalewaytasks/instance.go

```diff
@@ -0,0 +1,199 @@
+		CommercialType: fi.String(server.CommercialType),
+		Image:          s.Image,
+		Tags:           server.Tags,
+		UserData:       s.UserData,
+		Lifecycle:      s.Lifecycle,
+	}, nil
+}
+
+func (s *Instance) Run(c *fi.Context) error {
+	return fi.DefaultDeltaRunMethod(s, c)
+}
+
+func (_ *Instance) RenderScw(c *fi.Context, actual, expected, changes *Instance) error {
+	cloud := c.Cloud.(scaleway.ScwCloud)
+	instanceService := cloud.InstanceService()
+	zone := scw.Zone(fi.StringValue(expected.Zone))
+
+	userData, err := fi.ResourceAsBytes(*expected.UserData)
+	if err != nil {
+		return fmt.Errorf("error rendering instances: %w", err)
+	}
+
+	newInstanceCount := expected.Count
+	if actual != nil {
+		if expected.Count == actual.Count {
+			return nil
+		}
+		newInstanceCount = expected.Count - actual.Count
+	}
+
+	// If newInstanceCount > 0, we need to create new instances for this group
+	for i := 0; i < newInstanceCount; i++ {
+
+		// We create the instance
+		srv, err := instanceService.CreateServer(&instance.CreateServerRequest{
+			Zone:           zone,
+			Name:           fi.StringValue(expected.Name),
+			CommercialType: fi.StringValue(expected.CommercialType),
+			Image:          fi.StringValue(expected.Image),
+			Tags:           expected.Tags,
+		})
+		if err != nil {
+			return fmt.Errorf("error creating instance of group %q: %w", fi.StringValue(expected.Name), err)
+		}
+
+		// We wait for the instance to be ready
+		_, err = instanceService.WaitForServer(&instance.WaitForServerRequest{
+			ServerID: srv.Server.ID,
+			Zone:     zone,
+		})
+		if err != nil {
+			return fmt.Errorf("error waiting for instance %s of group %q: %w", srv.Server.ID, fi.StringValue(expected.Name), err)
+		}
+
+		// We load the cloud-init script in the instance user data
+		err = instanceService.SetServerUserData(&instance.SetServerUserDataRequest{
+			ServerID: srv.Server.ID,
+			Zone:     srv.Server.Zone,
+			Key:      "cloud-init",
+			Content:  bytes.NewBuffer(userData),
+		})
+		if err != nil {
+			return fmt.Errorf("error setting 'cloud-init' in user-data for instance %s of group %q: %w", srv.Server.ID, fi.StringValue(expected.Name), err)
+		}
+
+		// We start the instance
+		_, err = instanceService.ServerAction(&instance.ServerActionRequest{
+			Zone:     zone,
+			ServerID: srv.Server.ID,
+			Action:   instance.ServerActionPoweron,
+		})
+		if err != nil {
+			return fmt.Errorf("error powering on instance %s of group %q: %w", srv.Server.ID, fi.StringValue(expected.Name), err)
+		}
+
+		// We wait for the instance to be ready
+		_, err = instanceService.WaitForServer(&instance.WaitForServerRequest{
+			ServerID: srv.Server.ID,
+			Zone:     zone,
+		})
+		if err != nil {
+			return fmt.Errorf("error waiting for instance %s of group %q: %w", srv.Server.ID, fi.StringValue(expected.Name), err)
+		}
+	}
+
+	// If newInstanceCount < 0, we need to delete instances of this group
+	for i := 0; i > expected.Count; i-- {
```

Comment:
Should this loop start from `newInstanceCount`, otherwise it wouldn't do much?
```suggestion
	for i := newInstanceCount; i > expected.Count; i-- {
```

Reply:
there was indeed an error, but it's that the loop should stop at newInstanceCount like in the one where there are instances to add. If expected.Count is lower than actual.Count, newInstanceCount will have a negative value that indicated how many instances we have to remove.

Reply:
Should there be a change in the loop here to reflect ^^^?

Reply:
no worries. thanks.